deducing-this.md

```diff
@@ -241,9 +241,34 @@ In other words, `X` is *not* a template parameter.
 
 The type of the parameter will be deduced as if it were the first parameter of a non-member function template and the implicit object parameter was passed as the first argument.
 
+### Deduction To Derived and CRTP
+
+Since the deduction rules for the `this` parameter stay unchanged, we can get rid of some of the reasons for CRTP (Curiously Recurring Template Pattern).
+
+Consider the following example:
+
+```
+struct B {
+    template <typename This>
+    void do_stuff(This&& this self);
+
+    template <typename T>
+    friend friend_do_stuff(T&& self);
+};
+
+struct D : B { };
+
+D d;
+
+friend_do_stuff(d); // invokes B::friend_do_stuff<D&>
+d.do_stuff();       // invokes B::do_stuff<D&>, not B::do_stuff<B&>
+```
+
+This obviates most needs for CRTP, since parent classes can now provide functions which return derived types by the simple process of template argument deduction. Obtaining the derived type is now easy, and the need for arcane techniques for informing the base class of the derived are largely superflous. (They are still needed for policies, but not `boost::operators`, for instance).
```

Comment:
Do you think we should add an example implementation of some `boost::operators` stuff? Could make for a good example - particularly since there's a lot of complexity there that comes from having the pass the derived class type around through multiple layers.

Reply:
Hell yes.

Reply:
Actually, having read the Boost.Operators documentation, that's apparently not the best example - the complexity they have there is actually because apparently for some compilers, inheriting from multiple, empty base classes makes objects bigger. 

Iterator Facade might be better.

Reply:
> apparently for some compilers, inheriting from multiple, empty base classes makes objects bigger

That surprises me somewhat. I haven't had any issues with that for my `optional` and `expected` implementations which have many layers of private inheritance of empty base classes, including multiple inheritance.

Reply:
From cppreference, since c++11:
> Empty base optimization is required for StandardLayoutTypes in order to maintain the requirement that the pointer to a standard-layout object, converted using reinterpret_cast, points to its initial member, which is why one of the requirements for a standard layout type is "has no base classes with non-static data members and has no base classes of the same type as the first non-static data member".

The complexity is no longer needed.

Reply:
Yeah I didn't read the [description](http://www.boost.org/doc/libs/1_66_0/libs/utility/operators.htm#old_lib_note) closely enough to realize it predates C++11. My bad!